File: hex/minimap.py
import pygame
from hex.viewport import Viewport


class Minimap(Viewport):

    def __init__(self,
                 screen_x1,
                 screen_y1,
                 screen_width,
                 screen_height,
                 zoom_level,
                 map,
                 primary_viewport=None,
                 color=None):
        # Initialize with minimap-specific defaults
        super().__init__(
            screen_x1=screen_x1,
            screen_y1=screen_y1,
            screen_width=screen_width,
            screen_height=screen_height,
            zoom_level=zoom_level,
            map=map,
            minimap=None,  # Minimap doesn't have its own minimap
            can_zoom=False,  # Minimap typically doesn't zoom
            can_move=False,  # Minimap typically doesn't move
            color=color if color is not None else (0, 0, 0),
            primary_viewport=primary_viewport,
            is_minimap=True,
            is_primary=False
        )
        self.primary_viewport = primary_viewport

    def draw(self, grid=None):
        self.surface.fill(self.color)
        if grid is not None:
            self._draw_grid(grid)
        self.draw_map()
        self.draw_camera()

    def draw_camera(self):
        """Draw a rectangle showing the camera view of the primary viewport."""
        if self.primary_viewport is None:
            return
        x1, x2 = self.primary_viewport.cam.x1, self.primary_viewport.cam.x2
        y1, y2 = self.primary_viewport.cam.y1, self.primary_viewport.cam.y2
        x1, y1 = self.world_to_surface((x1, y1))
        x2, y2 = self.world_to_surface((x2, y2))
        x1 = max(0, x1)
        y1 = max(0, y1)
        x2 = min(self.screen_width, x2)
        y2 = min(self.screen_height, y2)
        width = x2 - x1
        height = y2 - y1
        pygame.draw.rect(self.surface, (255, 255, 255), (x1, y1, width, height), 1)

    def handle_mouse_drag(self, mouse_pos, prev_mouse_pos, mouse_down):
        """Handle mouse drag panning on the minimap to move the primary viewport."""
        if not mouse_down:
            return False
        
        pos = self.mouse_on_surface()
        if pos is None:
            return False
        
        surface_x1, surface_y1 = pos
        world_x1, world_y1 = self.surface_to_world((surface_x1, surface_y1))
        
        if self.primary_viewport is not None:
            self.primary_viewport.cam.change(center=(world_x1, world_y1), map=self.primary_viewport.map)
            self.primary_viewport.moved = True
            return True
        
        return False

File: hex-monogame/HexEngine/Minimap.cs
using System;
using Microsoft.Xna.Framework;
using Microsoft.Xna.Framework.Graphics;
using HexEngine.Maps;

namespace HexEngine;

public class Minimap
{
    public Viewport View { get; }

    public float ScreenX1 => View.ScreenX1;
    public float ScreenY1 => View.ScreenY1;
    public float ScreenWidth => View.ScreenWidth;
    public float ScreenHeight => View.ScreenHeight;
    public RenderTarget2D? RenderTarget => View.RenderTarget;
    public bool IsMinimap => View.IsMinimap;
    public bool IsPrimary => View.IsPrimary;

    private readonly Viewport? _primaryViewport;
    private static readonly InteractionState _emptyState = new();

    public Minimap(float screenX1, float screenY1, float screenWidth, float screenHeight,
                   float zoomLevel, GridMap map, Viewport? primaryViewport = null, Color? color = null)
    {
        _primaryViewport = primaryViewport;
        View = new Viewport(screenX1, screenY1, screenWidth, screenHeight, zoomLevel, map,
               canZoom: false, canMove: false, color: color ?? Colors.BLACK,
               primaryViewport: primaryViewport, isMinimap: true, isPrimary: false);
    }

    public void CreateRenderTarget(GraphicsDevice graphicsDevice)
        => View.CreateRenderTarget(graphicsDevice);

    public void Draw(GraphicsDevice graphicsDevice, PrimitiveDrawer drawer, MapRenderer renderer)
    {
        if (View.RenderTarget == null) return;

        graphicsDevice.SetRenderTarget(View.RenderTarget);
        graphicsDevice.Clear(View.BackColor);

        drawer.UpdateProjection((int)View.ScreenWidth, (int)View.ScreenHeight);

        renderer.Draw(drawer, View, _emptyState, false);
        DrawCamera(drawer);

        graphicsDevice.SetRenderTarget(null);
    }

    private void DrawCamera(PrimitiveDrawer drawer)
    {
        if (_primaryViewport == null) return;

        float pf = EngineConfig.PerspectiveFactor;
        float cx1 = _primaryViewport.Cam.X1;
        float cx2 = _primaryViewport.Cam.X2;
        float cy1 = _primaryViewport.Cam.Y1;
        float cy2 = _primaryViewport.Cam.Y2;

        if (pf > 0)
        {
            var pv = _primaryViewport;
            var worldTL = pv.SurfaceToWorld(new Vector2(pv.ScreenX1, pv.ScreenY1));
            var worldTR = pv.SurfaceToWorld(new Vector2(pv.ScreenX1 + pv.ScreenWidth, pv.ScreenY1));
            var worldBL = pv.SurfaceToWorld(new Vector2(pv.ScreenX1, pv.ScreenY1 + pv.ScreenHeight));
            var worldBR = pv.SurfaceToWorld(new Vector2(pv.ScreenX1 + pv.ScreenWidth, pv.ScreenY1 + pv.ScreenHeight));

            var sTL = View.WorldToSurface(worldTL);
            var sTR = View.WorldToSurface(worldTR);
            var sBL = View.WorldToSurface(worldBL);
            var sBR = View.WorldToSurface(worldBR);

            var points = new Vector2[] { sTL, sTR, sBR, sBL };
            for (int i = 0; i < points.Length; i++)
            {
                points[i] = new Vector2(
                    Math.Clamp(points[i].X, 1, View.ScreenWidth - 1),
                    Math.Clamp(points[i].Y, 1, View.ScreenHeight - 1));
            }

            drawer.DrawPolygonOutline(points, Colors.WHITE);
        }
        else
        {
            var topLeft = View.WorldToSurface(new Vector2(cx1, cy1));
            var bottomRight = View.WorldToSurface(new Vector2(cx2, cy2));

            float sx1 = Math.Max(1, topLeft.X);
            float sy1 = Math.Max(1, topLeft.Y);
            float sx2 = Math.Min(View.ScreenWidth - 1, bottomRight.X);
            float sy2 = Math.Min(View.ScreenHeight - 1, bottomRight.Y);

            float w = sx2 - sx1;
            float h = sy2 - sy1;

            if (w > 0 && h > 0)
                drawer.DrawRectOutline(sx1, sy1, w, h, Colors.WHITE);
        }
    }

    public void HandleDrag((int X, int Y) mousePos, bool mouseDown)
    {
        if (!mouseDown) return;
        if (!View.IsWithin(mousePos.X, mousePos.Y)) return;

        var worldPos = View.SurfaceToWorld(new Vector2(mousePos.X, mousePos.Y));

        if (_primaryViewport != null)
        {
            _primaryViewport.Cam.Change(center: worldPos);
            _primaryViewport.Moved = true;
        }
    }
}
